dbcon/mysql/ha_mcs_dml.cpp

```diff
@@ -1659,134 +1572,44 @@ int ha_mcs_impl_write_batch_row_(const uchar* buf, TABLE* table, cal_impl_if::ca
-                    uint32_t dataLength = 0;
-                    uintptr_t* dataptr;
-                    uchar* ucharptr;
-                    uint32_t colWidthInBytes;
-
-                    bool isBlob = ci.columnTypes[colpos].colDataType == CalpontSystemCatalog::BLOB;
-                    
-                    if (isBlob)
-                    {
-                        colWidthInBytes = ci.columnTypes[colpos].colWidth;
-                    }
-                    else
-                    {
-                        // For TEXT fields, MDB sets char_length to the maximum number that will fit in the number of bytes the 
-                        // defined TEXT length will fit in.
-                        // Ex: 
-                        // TEXT(25) will have char_length() == 255; 
-                        // TEXT(200) for latin_1 will have char_length() = 255
-                        // TEXT(200) for udf8mb4 will have char_length() = 65535
-                        // the length 200 multiplied by mbmaxlen (4) is > 255, so it needs 2 bytes for length. MDB sets to max(uint_16t)
-                        colWidthInBytes = table->field[colpos]->char_length();
-                    }
-                    
-                    if (colWidthInBytes < 256)
-                    {
-                        dataLength = *(uint8_t*) buf;
-                        buf++;
-                    }
-                    else if (colWidthInBytes < 65536)
-                    {
-                        dataLength = *(uint16_t*) buf;
-                        buf += 2;
-                    }
-                    else if (colWidthInBytes < 16777216)
-                    {
-                        dataLength = *(uint16_t*) buf;
-                        dataLength |= ((int) buf[2]) << 16;
-                        buf += 3;
-                    }
-                    else
-                    {
-                        dataLength = *(uint32_t*) buf;
-                        buf += 4;
-                    }
-
-                    // buf contains pointer to blob, for example:
-                    // (gdb) p (char*)*(uintptr_t*)buf
-                    // $43 = 0x7f68500c58f8 "hello world"
-                    dataptr = (uintptr_t*)buf;
-                    ucharptr = (uchar*)*dataptr;
-                    buf += sizeof(uintptr_t);
-
-                    if (isBlob)
-                    {
-                        for (uint32_t i = 0; i < dataLength; i++)
-                        {
-                            fprintf(ci.filePtr, "%02x", *(uint8_t*)ucharptr);
-                            ucharptr++;
+                            escape.assign(value.ptr(), value.length());
+                            boost::replace_all(escape, "\\", "\\\\");
+                            fprintf(ci.filePtr, "%c%.*s%c%c", ci.enclosed_by, (int)escape.length(),
+                                escape.c_str(), ci.enclosed_by, ci.delimiter);
```

Comment:
Double copy happens here:
- From value to escape on "assign()".
- To a new buffer inside boost::replace_all()

As discussed earlier in slack, I think we need a method in "ci" like this:
```
void fprintEscaped(const char *str, size_t length)
{
  fprintf(filePtr, "%c", enclosed_by);
  for (const char *end= str + length; str < end; str++)
  {
    if (*str == '\\')
      fprintf(filePtr, "%c", '\\');
    fprintf(filePtr, "%c", *str);
  }
  fprintf(filePtr, "%c%c", enclosed_by, delimiter);
}
```

For now, it can still escape only backslash as before.
But later we'll need to escape more characters, as cpimport currently fails when the string has characters 0x07, 0x0A and 0x11.

Reply:
Double-copy is much cheaper than calling fprintf(3) on each char. I suspect that the most optimal would be function something like that (I haven't tested it, just as an example):
```cpp
void fwriteEscaped(FILE* fp, const char* str, size_t len)
{
    fprintf(fp, "%c", '\'');
    const char* first_not_written = str;
    const char* s = str;
    for (const char *end = str + len; s < end; s++)
    {
        if (*s == '\\')
        {
            if (first_not_written != str)
            {
                fwrite(first_not_written, s - first_not_written, 1, fp);
            }
            fputc('\\', fp);
            fputc(*str, fp);
            first_not_written = s + 1;
        }
    }                                                                                                                                                                                    
                                                                                                                                                                                         
    if (first_not_written != str + len)
    {
        fwrite(first_not_written, str + len - first_not_written, 1, fp);                                                                                                                 
    }                                                                                                                                                                                    
    fprintf(fp, "%c%c", '\'', ',');
}
```

Reply:
Quick and dumb benchmark (each function called 100000 times):
```
fwriteEscaped(bs): 177271us
fprintEscaped(bs): 1099690us
stringEscaped(bs): 205665us
fwriteEscaped(nobs): 13391us
fprintEscaped(nobs): 567740us
stringEscaped(nobs): 22922us
fwriteEscaped(somebs): 31323us
fprintEscaped(somebs): 612498us
stringEscaped(somebs): 83735us
```
* `bs` — 255 `\`
* `nobs` — 255 `a`
* `somebs` — 255 `a`, every 17th char replaced with `\`

Reply:
Thanks. Good catch. ```fputc(*str, f)``` is much better than ```fprintf("%c", *str)```

Reply:
Anyway, in fact the less external function calls the faster the code. A naive `fputc`-version is the fastest only when the string contains only `\` (obviously). In more realistic case `nobs` it is slower than both `stringEscape` and `fwriteEscape`, and in `somebs` case it is about 15% faster than `stringEscape`, but it's still twice as slow as `fwriteEscape`

Reply:
I'm grateful for the important input but I'll let you step in and introduce this optimization on your own b/c it goes future then I wish to optimize with this patch.